doc: add operator-scope kb layout (#2982)
operator-framework/operator-sdk: website/content/en/docs/kubebuilder/operator-scope.md

# Operator Scope

## Overview

A namespace-scoped operator watches and manages resources in a single Namespace, whereas a cluster-scoped operator watches and manages resources cluster-wide.

A cluster-scoped scope is ideal for operator projects which will control resources in more than one Namespace. 
At the same time, namespace-scoped operators are preferred because of their flexibility. They enable decoupled upgrades, namespace isolation for failures and monitoring, and differing API definitions.

By default, `operator-sdk init` scaffolds a cluster-scoped operator. This document details conversion of default 
operator projects to namespaced-scoped operators. However, there are use cases where a cluster-scoped operator may make sense. For example, the [cert-manager][cert-manager] operator is often deployed with cluster-scoped permissions and watches so that it can manage and issue certificates for an entire cluster.

**IMPORTANT**: When a [Manager][ctrl-manager] instance is created in the `main.go` file, the 
Namespace(s) is set via [Manager Options][ctrl-options] as described below. These Namespace(s) should be watched and cached for the Client which is

> provided by the Controllers. Only clients provided by cluster-scoped projects are able to manage cluster-scoped CRD's. For further information see: [CRD scope doc][crd-scope-doc].

Author: it is true .. controller-runtime probably was the first thought here. Good catcher.

    LeaderElection:     enableLeaderElection, 
    LeaderElectionID:   "f1c5ece8.example.com",
})
...
```

**NOTE** The above example is the default code generated with `operator-sdk init` which initializes the Manager to be cluster-scoped.

## Watching resources in a single Namespace

To restrict the scope of the [Manager's][ctrl-manager] cache to a specific Namespace set the Namespace attribute from [Options][ctrl-options]:

```go
...
mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    Namespace:          "operator-namespace",
    MetricsBindAddress: metricsAddr,
})
...
``` 

## Watching resources in a set of Namespaces

It is possible to use the [MultiNamespacedCacheBuilder][multi-namespaced-cache-builder] attribute from 
[Options][ctrl-options] to watch and manage resources only in a specific set of Namespaces:

```go
...
namespaces := []string{"foo", "bar"} // List of Namespaces

mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    NewCache:           cache.MultiNamespacedCacheBuilder(namespaces),
    MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
})
...
```

> In the above example, if a CR(CustomResource), which is managed by its controllers, is applied to a Namespace that is not in this List of Namespaces then, its controller will not re-trigger its reconcile since the Namespaced is NOT part of the [Manager][ctrl-manager] scope.

Reviewer: ```suggestion
In the above example, a CR created in a Namespace not in the set passed to `Options` will not be reconciled by its controller because the Manager does not manage that Namespace.
```

> **IMPORTANT:** Note that this is not intended to be used for excluding Namespaces, this is better done via a Predicate. Also note that you may face performance issues when using this with a high number of Namespaces.

Reviewer: ```suggestion
**IMPORTANT:** Note that this is not intended to be used for excluding Namespaces, this is better done via a Predicate.
Also note that you may face performance issues when managing a high number of Namespaces.
```

## Restricting Roles and permissions

> The operator-scope defines the [Manager][ctrl-manager] cache scope but not the permissions to access the resources. After updating the Manager's scope to be Namespaced, the [Kubernetes Role-Based Access Control or the (RBAC)][k8s-rbac] permissions should be restricted accordingly.

Reviewer: ```suggestion
An operator's scope defines its [Manager's][ctrl-manager] cache's scope but not the permissions to access the resources. After updating the Manager's scope to be Namespaced, the cluster's [Role-Based Access Control (RBAC)][k8s-rbac] permissions should be restricted accordingly.
```

> These permissions are found in the directory `config/rbac/`. The `ClusterRole` and `ClusterRoleBinding` manifests are responsible for setup the permissions which allow the operator to have access to the resources.

Reviewer: ```suggestion
These permissions are found in the directory `config/rbac/`. The `ClusterRole` and `ClusterRoleBinding` manifests are responsible for creating the permissions which allow the operator to have access to the resources.
```

> **NOTE** The other files that can be found in this directory not requires changes to achieve this goal. The RBAC manifests with the names `<kind>_editor_role.yaml` and `<kind>_viewer_role.yaml` are not used by the operator and are meant to be helpers for specific config to allow access to users and clients per Kind. Also, the files with the names `auth_proxy_*.yaml` are specific to the [kube-auth-proxy][kube-rbac-proxy]. 

Reviewer: No need to explain what each RBAC manifest does. Leave that to other docs.
```suggestion
**NOTE** Only `ClusterRole` and `ClusterRoleBinding` manifests require changes to achieve this goal. Ignore `<kind>_editor_role.yaml`, `<kind>_viewer_role.yaml`, and files a name pattern of `auth_proxy_*.yaml`.
```

### Changing the permissions 

> To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use a `Role` instead of `ClusterRole` and `RoleBinding` instead of c

Reviewer: ```suggestion
To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use `Role`s and `RoleBinding`s instead of `ClusterRole`s and `ClusterRoleBinding`s, respectively.
```

> [`RBAC markers`][rbac-markers] defined in the controller (e.g `controllers/cache/memcached_controller.go`)

Reviewer: ```suggestion
[RBAC markers][rbac-markers] defined in the controller (e.g `controllers/memcached_controller.go`)
```

are used to generate the operator's [RBAC ClusterRole][rbac-clusterrole] (e.g `config/rbac/role.yaml`). The default markers don't specify a `namespace` property and will result in a ClusterRole.

Update the RBAC markers to specify a `namespace` property so that `config/rbac/role.yaml` is generated as a `Role` instead of a `ClusterRole`.

Replace:

```go
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds/status,verbs=get;update;patch
```

> With the Namespace `marker`:

Reviewer: ```suggestion
With namespaced markers:
```

```go
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds/status,verbs=get;update;patch
```

And then, run `make manifests` to update `config/rbac/role.yaml`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
...
```

> <!-- todo(camilamacedo86): remove the note or improve the above info when we have further info in the issue tracked  -->     

Reviewer: This may have been due to a typo: https://github.com/kubernetes-sigs/kubebuilder/issues/1496#issuecomment-631160626.

Author: unfortunately, it is not. Info added in the issue.

Reviewer: Since ^ is not a bug I recommend removing this section and instead say
```
We also need to update our ClusterRoleBindings to RoleBindings since they are not regenerated
```

Author: I still believe that it is a BUG but I will leave the comment and the step as suggested.

**NOTE**: Note that the above steps are not updating the `ClusterRoleBinding` to `RoleBinding` in `config/rbac/role_binding.yaml` which need to be done manually and shows a bug. More info: https://github.com/kubernetes-sigs/kubebuilder/issues/1496.

## Using environment variables for Namespace 

> Instead of having the specific Namespace(s) hard-coded in the `main.go` file a good practice is to use environment 

Reviewer: ```suggestion
Instead of having any Namespaces hard-coded in the `main.go` file a good practice is to use environment 
```

variables to allow the restrictive configurations

### Configuring Namespace scoped operators

- Add a helper function in the `main.go` file:

```go
// getWatchNamespace returns the Namespace the operator should be watching for changes

> func getWatchNamespace() (string, error) {

Reviewer: Better to recommend [`k8sutil.GetWatchNamespace()`](https://github.com/operator-framework/operator-sdk/blob/master/pkg/k8sutil/k8sutil.go#L46).

Author: We have not k8sutil in KB layout.

Reviewer: `k8sutil` is a library that is meant to be imported. I'm not aware of this library being removed. I'd rather see a user use library code, although this function is pretty straightforward.

Author: We have no instructions in the quick start and in any other places telling for users add SDK module as dep lib and we do not know what we will or not port from there to KB layout. So, I do not think that we should add it here. Is it make sense? 

c/c @hasbro17 @joelanford wdyt?

mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    MetricsBindAddress: metricsAddr,
    Port:               9443,
    LeaderElection:     enableLeaderElection,
    LeaderElectionID:   "f1c5ece8.example.com",
    Namespace:          watchNamespace, // namespaced-scope when the value is not an empty string 
})
...
```

- Define the environment variable in the `config/manager/manager.yaml`:

```yaml
spec:
  containers:
  - command:
    - /manager
    args:
    - --enable-leader-election
    image: controller:latest
    name: manager
    resources:
      limits:
        cpu: 100m
        memory: 30Mi
      requests:
        cpu: 100m
        memory: 20Mi
    env:
      - name: WATCH_NAMESPACE
        valueFrom:
          fieldRef:
            fieldPath: metadata.namespace 
  terminationGracePeriodSeconds: 10
``` 

**NOTE** The above will set as value of `WATCH_NAMESPACE` the namespace where the operator will be deployed. 

### Configuring cluster-scoped operators with MultiNamespacedCacheBuilder

- Add a helper function in the `main.go` file :

```go
// getWatchNamespace returns the namespace the operator should be watching for changes

> func getWatchNamespace() (string, error) {

Reviewer: Ditto above.

Author: ^ditto

Author: We have not k8sutil in KB layout and we do not know if we will keep it.

    Namespace:          watchNamespace, // namespaced-scope when the value is not an empty string 
})
...
```

- Define the environment variable in the `config/manager/manager.yaml`:

```yaml
spec:
  containers:
  - command:
    - /manager
    args:
    - --enable-leader-election
    image: controller:latest
    name: manager
    resources:
      limits:
        cpu: 100m
        memory: 30Mi
      requests:
        cpu: 100m
        memory: 20Mi
    env:
      - name: WATCH_NAMESPACE
        valueFrom:
          fieldRef:
            fieldPath: metadata.namespace 
  terminationGracePeriodSeconds: 10
``` 

**NOTE** `WATCH_NAMESPACE` here will always be set as the namespace where the operator is deployed. 

### Configuring cluster-scoped operators with MultiNamespacedCacheBuilder

- Add a helper function to get the environment variable value in the `main.go` file as done in the in the previous example (e.g `getWatchNamespace()`) 
- Use the environment variable value and check if it is a multi-namespace scenario:

```go
    ...
watchNamespace, err := getWatchNamespace()
if err != nil {
    setupLog.Error(err, "unable to get WatchNamespace, " +

>         "the manager will watch and manage resources in all cluster")

Reviewer 2: ```suggestion
        "the manager will watch and manage resources in all namespaces")
```

## Restricting Roles and permissions

The operator-scope defines the [Manager][ctrl-manager] cache scope but not the permissions to access the resources. After updating the Manager's scope to be Namespaced, the [Kubernetes Role-Based Access Control or the (RBAC)][k8s-rbac] permissions should be restricted accordingly.

These permissions are found in the directory `config/rbac/`. The `ClusterRole` and `ClusterRoleBinding` manifests are responsible for setup the permissions which allow the operator to have access to the resources.

**NOTE** The other files that can be found in this directory not requires changes to achieve this goal. The RBAC manifests with the names `<kind>_editor_role.yaml` and `<kind>_viewer_role.yaml` are not used by the operator and are meant to be helpers for specific config to allow access to users and clients per Kind. Also, the files with the names `auth_proxy_*.yaml` are specific to the [kube-auth-proxy][kube-rbac-proxy]. 

### Changing the permissions 

> To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use a `Role` instead of `ClusterRole` and `RoleBinding` instead of c

Reviewer 2: @camilamacedo86 I think you missed this from eric's last review
```suggestion
To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use `Role`s and `RoleBinding`s instead of `ClusterRole`s and `ClusterRoleBinding`s, respectively.
```

> # Operator Scope

Author: It is the new doc.

## Overview

> A `namespace-scoped` operator watches and manages resources in a single namespace, whereas a `cluster-scoped` operator watches and manages resources cluster-wide. 

Reviewer: The *-scoped formatting doesn't make sense, since those terms are not formal objects:
```suggestion
A namespace-scoped operator watches and manages resources in a single namespace, whereas a cluster-scoped operator watches and manages resources cluster-wide. 
```

> A `namespace-scoped` operator watches and manages resources in a single namespace, whereas a `cluster-scoped` operator watches and manages resources cluster-wide. 

Reviewer: Meta note: we should consider writing documents with short-line paragraphs (100 characters or less per line) for new or re-formatted documents. They will be easier to review. This has been suggested a few times already.

Reviewer 3: I overlook the long lines, but they are a pain to read IMO :)

A `cluster-scoped` scope is ideal for operator projects which will control resources in more than one namespace. At the same time, `namespace-scoped` operators are preferred because of their flexibility. They enable decoupled upgrades, namespace isolation for failures and monitoring, and differing API definitions.

> However, there are use cases where a `cluster-scoped` operator may make sense. For example, the [cert-manager][cert-manager] operator is often deployed with cluster-scoped permissions and watches so that it can manage to issue certificates for an entire cluster.

Reviewer 3: change: manage to issue -> manage and issue

> However, there are use cases where a `cluster-scoped` operator may make sense. For example, the [cert-manager][cert-manager] operator is often deployed with cluster-scoped permissions and watches so that it can manage to issue certificates for an entire cluster.

Reviewer: This sentence doesn't really follow well from the previous paragraph. Something along the lines of the following would be better:
```
By default, `operator-sdk init` scaffolds a cluster-scoped operator.
This document details conversion of default operator projects to namespaced-scoped operators.
```

> **IMPORTANT**: When a [Manager][ctrl-manager] instance is created in the `main.go` file, it might receives the namespace(s) as Options as described bellow. These namespace(s) should be watched and cached for the Client which is provided by the Controllers. Only clients provided by `cluster-scoped` projects are able to manage `cluster-scoped` CRD's. For further information see: [CRD scope doc][crd-scope-doc].

Reviewer 3: typo: bellow -> below

## Watching and Managing resources in all cluster (Default)

> An [Manager][ctrl-manager] is responsible for manage the controllers of the operator project. In the `main.go` you will be able to find it. See:

Reviewer 3: typo: An -> A

Reviewer 3: typo: manage the controllers -> managing the controllers

```go
    ...
	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
		Scheme:             scheme,
		MetricsBindAddress: metricsAddr,
		Port:               9443,
		LeaderElection:     enableLeaderElection,
		LeaderElectionID:   "f1c5ece8.example.com",
	})
	if err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}
    ...
```

> **NOTE** The above example is the default code generated for the projects created with `operator-sdk init` which are creating a new Manager for an operator `cluster-scoped`.

Reviewer 3: typo: are creating -> is creating

## Watching and Managing resources in an Namespace

To restrict the scope of the [Manager's][ctrl-manager] cache to an specific `namespace` is required to inform it via the `Namespace` attribute from [Options][ctrl-options]. See: 

```go
    ...
    mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
        Scheme:             scheme,
        Namespace:          "operator-namespace",
        MetricsBindAddress: metricsAddr,
    })
    ...
``` 

## Watching and Managing resources in a set of Namespaces

It is possible to use the [MultiNamespacedCacheBuilder][multi-namespaced-cache-builder] attribute from [Options][ctrl-options] to watch and manage resources only in a specific set of namespaces. See:

```go
    ...
    var namespaces []string // List of Namespaces

    mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
        Scheme:             scheme,
        NewCache:           cache.MultiNamespacedCacheBuilder(namespaces),
        MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
    })
    ...
```

> In the above example, if a CR(CustomResource) which is managed by its controllers be applied in a namespace that is not in this List of `namespaces` then, its controller will not re-trigger its reconcile since the namespaced is NOT part of the [Manager][ctrl-manager] scope.

Reviewer 3: rephrase: if a CR(CustomResource) which is managed by its controllers be applied in a namespace. ->
if a CR(CustomResource), which is managed by its controllers, is applied to a namespace

-  In `config/rbac/kustomization.yaml`:
```yaml
...
# Comment the following 4 lines if you want to disable
# the auth proxy (https://github.com/brancz/kube-rbac-proxy)
# which protects your /metrics endpoint.
- auth_proxy_service.yaml
- auth_proxy_role.yaml
- auth_proxy_role_binding.yaml
- auth_proxy_client_clusterrole.yaml
```

- In `config/default/kustomization.yaml`:
```yaml
...
patchesStrategicMerge:
  # Protect the /metrics endpoint by putting it behind auth.
  # If you want your controller-manager to expose the /metrics
  # endpoint w/o any authn/z, please comment the following line.
- manager_auth_proxy_patch.yaml
...
```

And then, see that for each API (CRD) scaffolded a specific role file with permissions to `edit` or `view` the Kind resource in the cluster is created too. It is not used by the project, but are helpers to allow you to perform specific configurations. For example, after creating a new API(CRD) with the Memcached Kind, we will able to see the roles `memcached_editor_role.ymal` and `memcached_viewer_role.yaml` created there.

### Changing the permissions 

> To change the scope of the RBCA permissions from cluster-wide to a specific namespace, you will need to use a `Role` instead of `ClusterRole` and `RoleBinding` instead of `ClusterRoleBinding`. Note that, the project uses `markers` to define the properties of these RBCA files which are generated. The [RBAC ClusterRole][rbca-clusterrole] are created by default because of the `markers` in the Kind's controllers are:

Reviewer 3: typo: RBCA -> RBAC (2x)

```yaml
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds/status,verbs=get;update;patch
```

In this way, it is required to add the namespace `marker`. See:

```yaml
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds/status,verbs=get;update;patch
```

> Then, run the command `make manifests` to updated the files generated. 

Reviewer 3: rephrase: to updated the files generated -> to update the generated files.

<!-- todo(camilamacedo86): remove the note or improve the above info when we have further info in the issue tracked  -->     
**NOTE**: Note that the above steps are currently not changing the all required files and it shows a bug which is tracked in the Kubebuilder repository. More info: https://github.com/kubernetes-sigs/kubebuilder/issues/1496

## Using environment variables for namespace 

> Instead of having the specific namespace(s) harder-coded in the `main.go` file a good practice is to use environment variables to allow the restrictive configurations

Reviewer 3: typo: harder-coded -> hard-coded

namespace isolation for failures and monitoring, and differing API definitions.

By default, `operator-sdk init` scaffolds a cluster-scoped operator. This document details conversion of default 
operator projects to namespaced-scoped operators. However, there are use cases where a cluster-scoped operator may 
make sense. For example, the [cert-manager][cert-manager] operator is often deployed with cluster-scoped permissions 
and watches so that it can manage and issue certificates for an entire cluster.

**IMPORTANT**: When a [Manager][ctrl-manager] instance is created in the `main.go` file, the 
namespace(s) is set via [Manager Options][ctrl-options] as described below. These namespace(s) should be watched and cached for the Client which is
provided by the Controllers. Only clients provided by cluster-scoped projects are able to manage cluster-scoped CRD's. 
For further information see: [CRD scope doc][crd-scope-doc].

## Watching resources in all Namespaces (default)

A [Manager][ctrl-manager] is initialized with no Namespace option specified, or `Namespace: ""` will watch all Namespaces. In the `main.go` you will 

> be able to find it. See:

Author: ok make sense.

```go
...
namespaces := []string{"foo", "bar"} // List of Namespaces

mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    NewCache:           cache.MultiNamespacedCacheBuilder(namespaces),
    MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
})
...
```

In the above example, if a CR(CustomResource), which is managed by its controllers, is applied to a namespace that 
is not in this List of `namespaces` then, its controller will not re-trigger its reconcile since the namespaced is NOT 
part of the [Manager][ctrl-manager] scope.

**IMPORTANT:** Note that this is not intended to be used for excluding namespaces, this is better done via a Predicate. 
Also note that you may face performance issues when using this with a high number of namespaces.

## Restricting Roles and permissions

The operator-scope defines the [Manager][ctrl-manager] cache scope but not the permissions to access the resources. After updating the Manager's scope to be Namespaced, the [Kubernetes Role-Based Access Control or the (RBAC)][k8s-rbac] permissions should be restricted accordingly.

> In the projects, the permissions are specified in the directory `config/rbac/`. The RBAC manifest at `config/rbac/role.yaml` and `config/rbac/role_binding.yaml` specifies the permissions needed by the operator to manage its resources, and is a `ClusterRole` by default. 

Reviewer 2: The first sentence is redundant. We're already highlighting that the RBAC manifests are in `config/rbac/...` in the next one.

Also the rolebinding manifest does not specify permissions. But if you want to mention `rolebinding.yaml` here then you need to rephrase the sentence a bit.
```suggestion
The `ClusterRole` manifest at `config/rbac/role.yaml` specifies the permissions needed by the operator to manage its resources. 
The `ClusterRoleBinding` manifest at `config/rbac/role_binding.yaml` binds that to the operator's `ServiceAccount`.
```

Author: Addressed your comment by trying to add an smaller info yet.

 
**NOTE** The files `auth_proxy_client_clusterrole.yaml`, `auth_proxy_role.yaml`, `auth_proxy_role_binding.yaml` are also 
defined to allow permissions across the cluster as well and are used by the operator. More info see: 
[kube-rbac-proxy][kube-rbac-proxy]. Also, it is important to highlight that it is an alpha feature that is not officially 
supported and can be disabled by commenting the lines in the config `config/rbac/kustomization.yaml` and in the 
`config/default/kustomization.yaml`. See:

-  In `config/rbac/kustomization.yaml`:
```yaml
...
# Comment the following 4 lines if you want to disable
# the auth proxy (https://github.com/brancz/kube-rbac-proxy)
# which protects your /metrics endpoint.
- auth_proxy_service.yaml
- auth_proxy_role.yaml
- auth_proxy_role_binding.yaml
- auth_proxy_client_clusterrole.yaml
```

- In `config/default/kustomization.yaml`:
```yaml
...
patchesStrategicMerge:
  # Protect the /metrics endpoint by putting it behind auth.
  # If you want your controller-manager to expose the /metrics
  # endpoint w/o any authn/z, please comment the following line.
- manager_auth_proxy_patch.yaml
...
```

> And then, see that for each API (CRD) scaffolded a specific role file with permissions to `edit` or `view` the Kind 
> resource in the cluster is created too. It is not used by the project, but are helpers to allow you to perform specific 
> configurations. For example, after creating a new API(CRD) with the Memcached Kind, we will able to see the roles 
> `memcached_editor_role.ymal` and `memcached_viewer_role.yaml` created there.

Reviewer 2: Grammar and typos. 
Also we can rephrase this to be a lot more concise as well.
```suggestion
The RBAC manifests with the names `<kind>_editor_role.yaml` and `<kind>_viewer_role.yaml` are meant for end users or clients to view and edit CustomResources for that Kind.
```

Author: Addressed with small changes.

### Changing the permissions 

> To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use a `Role` 
> instead of `ClusterRole` and `RoleBinding` instead of `ClusterRoleBinding`. Note that, the project uses `markers` to 
> define the properties of these RBAC files which are generated. The [RBAC ClusterRole][rbac-clusterrole] are created by 
> default because of the `markers` in the Kind's controllers are:

Reviewer 2: - Remove extra `,`.
- Link to [RBAC markers](https://book.kubebuilder.io/reference/markers/rbac.html) since we're mentioning them here for the first time.
- Highlight why default is ClusterRole.
- Also rephrase since markers don't "define the properties of all files". Just the `role.yaml` manifest.
```suggestion
To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use a `Role` 
instead of `ClusterRole` and `RoleBinding` instead of `ClusterRoleBinding`. 

[`RBAC markers`][rbac-markers] defined in the controller (e.g `controllers/cache/memcached_controller.go`)
are used to generate the operator's `Role`/`ClusterRole` manifest `config/rbac/role.yaml`.
The default markers don't specify a `namespace` property and will result in a ClusterRole.

[rbac-markers]: https://book.kubebuilder.io/reference/markers/rbac.html
```

Author: Addressed with caveats.

```yaml
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds/status,verbs=get;update;patch
```

> In this way, it is required to add the namespace `marker`. See:

Reviewer 2: Mention exactly what we need to change in the marker to make the change from `ClusterRole` to `Role`:
```suggestion
Update the RBAC markers to specify a `namespace` property so that `config/rbac/role.yaml` is generated as a `Role` instead of a `ClusterRole`:
```

```yaml
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,namespace="my-namespace",resources=memcacheds/status,verbs=get;update;patch
```

> Then, run the command `make manifests` to update the generated files. 

Reviewer 2: No need for `,`:
```suggestion
Run `make manifests` to update `config/rbac/role.yaml`:
```

Author: It is the second step then requires `And then, xxxxx`. 
Also, regards to replace `.` with `:` done and added an example so.

[Options][ctrl-options] to watch and manage resources in a set of Namespaces:

```go
...
namespaces := []string{"foo", "bar"} // List of Namespaces

mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    NewCache:           cache.MultiNamespacedCacheBuilder(namespaces),
    MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
})
...
```
In the above example, a CR created in a Namespace not in the set passed to `Options` will not be reconciled by 
its controller because the [Manager][ctrl-manager] does not manage that Namespace.

**IMPORTANT:** Note that this is not intended to be used for excluding Namespaces, this is better done via a Predicate. 
Also note that you may face performance issues when managing a high number of Namespaces.

## Restricting Roles and permissions

An operator's scope defines its [Manager's][ctrl-manager] cache's scope but not the permissions to access the resources. 
After updating the Manager's scope to be Namespaced, the cluster's [Role-Based Access Control (RBAC)][k8s-rbac] 
permissions should be restricted accordingly.

> These permissions are found in the directory `config/rbac/`. The `ClusterRole` and `ClusterRoleBinding` manifests are 
> responsible for creating the permissions which allow the operator to have access to the resources.

Reviewer 2: `The ClusterRole and ClusterRoleBinding manifests` isn't specific enough because the other manifests are also ClusterRoles.
```suggestion
These permissions are found in the directory `config/rbac/`. The `ClusterRole` in `role.yaml` and `ClusterRoleBinding` in `role_binding.yaml` are 
used to grant the operator permissions to access and manage its resources.
```

> **NOTE** Only `ClusterRole` and `ClusterRoleBinding` manifests require changes to achieve this goal. 
> Ignore `<kind>_editor_role.yaml`, `<kind>_viewer_role.yaml`, and files a name pattern of `auth_proxy_*.yaml`.

Reviewer 2: - `Only ClusterRole and ClusterRoleBinding manifests require changes` is incorrect.
All manifests are ClusterRoles. We need to be specific about the manifest file names.
- Also mention why we ignore the other ones for now.
- Grammar nit.
```suggestion
**NOTE** For changing the operator's scope only the `role.yaml` and `role_binding.yaml` manifests need to be updated.
For the purposes of this doc, the other RBAC manifests `<kind>_editor_role.yaml`, `<kind>_viewer_role.yaml`, and `auth_proxy_*.yaml` are not relevant to changing the operator's resource permissions.
```

### Changing the permissions 

To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use `Role`s 
and `RoleBinding`s instead of `ClusterRole`s and `ClusterRoleBinding`s, respectively.
 
[`RBAC markers`][rbac-markers] defined in the controller (e.g `controllers/memcached_controller.go`)
are used to generate the operator's [RBAC ClusterRole][rbac-clusterrole] (e.g `config/rbac/role.yaml`). The default

>  markers don't specify a `namespace` property and will result in a ClusterRole.

Reviewer 2: ```suggestion
 markers don't specify a `namespace` property and will result in a `ClusterRole`.
```

```

And then, run `make manifests` to update `config/rbac/role.yaml`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
...
```

We also need to update our `ClusterRoleBindings` to `RoleBindings` since they are not regenerated: 

```yaml
  
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: manager-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: manager-role
subjects:
- kind: ServiceAccount
  name: default
  namespace: system
```

<!-- todo(camilamacedo86): The need for the RoleBinding show an issue tracked 
in https://github.com/kubernetes-sigs/kubebuilder/issues/1496. --> 

## Using environment variables for Namespace 

Instead of having any Namespaces hard-coded in the `main.go` file a good practice is to use environment 
variables to allow the restrictive configurations

### Configuring Namespace scoped operators

- Add a helper function in the `main.go` file:

```go
// getWatchNamespace returns the Namespace the operator should be watching for changes
func getWatchNamespace() (string, error) {
    // WatchNamespaceEnvVar is the constant for env variable WATCH_NAMESPACE

>     // which is the Namespace where the watch activity happens.
>     // this value is empty if the operator is running with clusterScope.

Reviewer 2: I realize this is straight from our util but `watch activity happens` is awkward phrasing.
```suggestion
    // which specifies the Namespace to watch.
    // An empty value means the operator is running with cluster scope.
```

    var watchNamespaceEnvVar = "WATCH_NAMESPACE"
    
    ns, found := os.LookupEnv(watchNamespaceEnvVar)
    if !found {
        return "", fmt.Errorf("%s must be set", watchNamespaceEnvVar)
    }
    return ns, nil
}
```

- Use the environment variable value: 

```go
...
watchNamespace, err := getWatchNamespace()
if err != nil {
    setupLog.Error(err, "unable to get WatchNamespace, " +

>         "the manager will watch and manage resources in all cluster")

Reviewer 2: There's only 1 cluster.
```suggestion
        "the manager will watch and manage resources in all namespaces")
```

}

mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    MetricsBindAddress: metricsAddr,
    Port:               9443,
    LeaderElection:     enableLeaderElection,
    LeaderElectionID:   "f1c5ece8.example.com",
    Namespace:          watchNamespace, // namespaced-scope when the value is not an empty string 
})
...
```

- Define the environment variable in the `config/manager/manager.yaml`:

```yaml
spec:
  containers:
  - command:
    - /manager
    args:
    - --enable-leader-election
    image: controller:latest
    name: manager
    resources:
      limits:
        cpu: 100m
        memory: 30Mi
      requests:
        cpu: 100m
        memory: 20Mi
    env:
      - name: WATCH_NAMESPACE
        valueFrom:
          fieldRef:
            fieldPath: metadata.namespace 
  terminationGracePeriodSeconds: 10
``` 

> **NOTE** The above will set as value of `WATCH_NAMESPACE` the namespace where the operator will be deployed. 

Reviewer 2: Grammar nit:
```suggestion
**NOTE**  `WATCH_NAMESPACE` here will always be set as the namespace where the operator is deployed. 
```

### Configuring cluster-scoped operators with MultiNamespacedCacheBuilder

- Add a helper function in the `main.go` file :

```go
// getWatchNamespace returns the namespace the operator should be watching for changes
func getWatchNamespace() (string, error) {
    // WatchNamespaceEnvVar is the constant for env variable WATCH_NAMESPACE
    // which is the namespace where the watch activity happens.
    // this value is empty if the operator is running with clusterScope.
    var watchNamespaceEnvVar = "WATCH_NAMESPACE"
	
    ns, found := os.LookupEnv(watchNamespaceEnvVar)
    if !found {
        return "", fmt.Errorf("%s must be set", watchNamespaceEnvVar)
    }
    return ns, nil
}
```

> - Use the environment variable value and check if it is an multi-namespace scenario:

Reviewer 2: Grammar nit:
```suggestion
- Use the environment variable value and check if it is a multi-namespace scenario:
```

```go
    ...
watchNamespace, err := getWatchNamespace()
if err != nil {
    setupLog.Error(err, "unable to get WatchNamespace, " +

>         "the manager will watch and manage resources in all cluster")

Reviewer 2: We're pretty much repeating the helper and main.go almost exactly as it is from the previous section so maybe remove that and just reference that above and just show `options.NewCache = cache.MultiNamespacedCacheBuilder`.

Or if not then make the same fixes to the helper code as I've pointed above.

# Operator Scope

## Overview

> A namespace-scoped operator watches and manages resources in a single namespace, whereas a cluster-scoped operator 

Reviewer: Can you break up lines in this doc, ex:
```suggestion
A namespace-scoped operator watches and manages resources in a single namespace,
whereas a cluster-scoped operator watches and manages resources cluster-wide.
```

Author: Hi @estroz, 

The IntelliJ shows a line with the expect line size/lint. All doc is respecting that so I cannot see a reason to do this break. See:

<img width="948" alt="Screenshot 2020-05-12 at 11 55 15" src="https://user-images.githubusercontent.com/7708031/81676320-9d3f9480-9447-11ea-8828-c66a8ba3d477.png">

watches and manages resources cluster-wide.

A cluster-scoped scope is ideal for operator projects which will control resources in more than one namespace. 
At the same time, namespace-scoped operators are preferred because of their flexibility. They enable decoupled upgrades, 
namespace isolation for failures and monitoring, and differing API definitions.

By default, `operator-sdk init` scaffolds a cluster-scoped operator. This document details conversion of default 
operator projects to namespaced-scoped operators. However, there are use cases where a cluster-scoped operator may 
make sense. For example, the [cert-manager][cert-manager] operator is often deployed with cluster-scoped permissions 
and watches so that it can manage and issue certificates for an entire cluster.

**IMPORTANT**: When a [Manager][ctrl-manager] instance is created in the `main.go` file, it might receives the 
namespace(s) as Options as described below. These namespace(s) should be watched and cached for the Client which is 
provided by the Controllers. Only clients provided by cluster-scoped projects are able to manage cluster-scoped CRD's. 
For further information see: [CRD scope doc][crd-scope-doc].

> ## Watching and Managing resources in all cluster (Default)

Author: Since namespace is a Type as Deployment, for example, I will keep just it capitalize

A [Manager][ctrl-manager] is responsible for managing the controllers of the operator project. In the `main.go` you will 

> be able to find it. See:

Author: Addressed the suggestion but keep the code example to illustrate what and where to help the users.

```go
...
mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    MetricsBindAddress: metricsAddr,
    Port:               9443,
    LeaderElection:     enableLeaderElection, 
    LeaderElectionID:   "f1c5ece8.example.com",
})
if err != nil {
    setupLog.Error(err, "unable to start manager")
    os.Exit(1)
}
...
```

**NOTE** The above example is the default code generated for the projects created with `operator-sdk init` which is 

> creating a new Manager for an operator `cluster-scoped`.

Author: Addressed suggestion. PS the `"cluster-scoped"` is defined and explained in the first section of this doc.

> ## Watching and Managing resources in an Namespace

Author: Addressed but keep Namespace since it is a type.

> To restrict the scope of the [Manager's][ctrl-manager] cache to an specific `namespace` is required to inform it via 
> the `Namespace` attribute from [Options][ctrl-options]. See: 

Reviewer 2: Why highlight "namespace"? We're not doing it anywhere else.
Plus we've already linked to the Manager godocs enough times above.
```suggestion
To restrict the scope of the Manager's cache to a specific namespace set the Namespace option:
```

Author: addressed with a few caveats.

```go
...
mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    Namespace:          "operator-namespace",
    MetricsBindAddress: metricsAddr,
})
...
``` 

> ## Watching and Managing resources in a set of Namespaces

Reviewer 2: ```suggestion
## Watching resources in a set of namespaces
```

It is possible to use the [MultiNamespacedCacheBuilder][multi-namespaced-cache-builder] attribute from 

> [Options][ctrl-options] to watch and manage resources only in a specific set of namespaces. See:

Reviewer 2: ```suggestion
[Options][ctrl-options] to watch and manage resources only in a specific set of namespaces:
```

```go
...

> var namespaces []string // List of Namespaces

Reviewer 2: Nit: Initialize to better illustrate the set.
```suggestion
namespaces := []string{"foo", "bar"} // List of Namespaces
```

mgr, err = ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:             scheme,
    NewCache:           cache.MultiNamespacedCacheBuilder(namespaces),
    MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
})
...
```

In the above example, if a CR(CustomResource), which is managed by its controllers, is applied to a namespace that 
is not in this List of `namespaces` then, its controller will not re-trigger its reconcile since the namespaced is NOT 
part of the [Manager][ctrl-manager] scope.

**IMPORTANT:** Note that this is not intended to be used for excluding namespaces, this is better done via a Predicate. 
Also note that you may face performance issues when using this with a high number of namespaces.

## Restricting Roles and permissions

The `operator-scope` defines the [Manager][ctrl-manager] cache scope but not the permissions to access the resources. 
In this scenario, it would also be recommended to change the default 

> [Kubernetes Role-Based Access Control or the (RBAC)][k8s-rbac] to be restrictive accordingly.

Author: > Why is the term operator-scope highlighted here?
Addressed. 

Note that it is a very common misunderstanding raised in channel and issues usually. So, IHMO it is important to highlight that the `[Manager][ctrl-manager] cache ` is not `(RBAC)][k8s-rbac]` as what is each one responsibility.

In the projects, the permissions are specified in the directory `config/rbac/`. Note that there you will find `role.yaml` 

> and `role_binding.yaml` files which are to allow permissions across the cluster by default for the operator. 

Author: Addressed but with few changes since the idea here is lead the user to the `rbac` dir since they will find more files than just the role one which are relevant to these scenarios.

> and `role_binding.yaml` files which are to allow permissions across the cluster by default for the operator. 

Author: Also, we need to speak about RoleBinding as well.

supported and can be disabled by commenting the lines in the config `config/rbac/kustomization.yaml` and in the 
`config/default/kustomization.yaml`. See:

-  In `config/rbac/kustomization.yaml`:
```yaml
...
# Comment the following 4 lines if you want to disable
# the auth proxy (https://github.com/brancz/kube-rbac-proxy)
# which protects your /metrics endpoint.
- auth_proxy_service.yaml
- auth_proxy_role.yaml
- auth_proxy_role_binding.yaml
- auth_proxy_client_clusterrole.yaml
```

- In `config/default/kustomization.yaml`:
```yaml
...
patchesStrategicMerge:
  # Protect the /metrics endpoint by putting it behind auth.
  # If you want your controller-manager to expose the /metrics
  # endpoint w/o any authn/z, please comment the following line.
- manager_auth_proxy_patch.yaml
...
```

And then, see that for each API (CRD) scaffolded a specific role file with permissions to `edit` or `view` the Kind 
resource in the cluster is created too. It is not used by the project, but are helpers to allow you to perform specific 
configurations. For example, after creating a new API(CRD) with the Memcached Kind, we will able to see the roles 

> `memcached_editor_role.ymal` and `memcached_viewer_role.yaml` created there.

Reviewer 2: @camilamacedo86 While I agree it is important to point all this out, this is probably not the right section. 
This part is derailing the flow of the doc where I only really care about the steps for changing `role.yaml` and `rolebinding.yaml`.

Let's keep this section strictly for the changes needed to change the operator's scope.
And move this to a caveats section at the end of this doc that points out what all the other rbac manifests are and why we might want to disable kube-rbac proxy.

With this, you can probably combine this section `Restricting Roles and permissions` and  `Changing the permissions` into one which is more concise and directly related to the permissions for the operator's scope.

@joelanford I know you had opinions on whether we should be documenting how to disable kube-rbac proxy downstream at all.

Author: Hi @hasbro17, 

Note that we need to make these notes here since: 
- Users might need to disable or change the proxy roles to achieve the goal. So, a small note as this without too many details allow they know where and how.  
- When users will be following these steps will check all these permissions files, so common questions will be raised such as: Has been these permissions also been applied and etc?. Then, IHMO it is important to keep a NOTE here.

Reviewer 2: @camilamacedo86 This note is hardly small. It's a complete section that diverts me into reading about kube-rbac proxy and how to disable it.

My point is that it's not necessary to modify the other ClusterRole/ClusterRolebinding manifests in `config/rbac` besides `role.yaml` and `rolebinding.yaml` just to change the permissions for a namespace scoped operator.
Please clarify if it is necessary.

So I still say we have to move this section to the end of this doc into a caveats section.

Author: Addressed as suggested.

### Changing the permissions 

To change the scope of the RBAC permissions from cluster-wide to a specific namespace, you will need to use `Role`s and `RoleBinding`s instead of `ClusterRole`s and `ClusterRoleBinding`s, respectively.
 

> [`RBAC markers`][rbac-markers] defined in the controller (e.g `controllers/cache/memcached_controller.go`)

Author: ```suggestion
[RBAC markers][rbac-markers] defined in the controller (e.g `controllers/memcached_controller.go`)
```
